daf_butler.py

```diff
@@ -0,0 +1,942 @@
+    butler defaults or a separate single data ID, which should be accepted by
+    any method accepting an expression of this type.
+
+    In queries that span multiple collections, dataset and quantum identifiers
+    that use data IDs are resolved in the order they appear in the collection
+    sequence, but range expressions can pick up datasets and quanta that are
+    shadowed.
+    """
+
+
+class QuantumStatus(enum.Enum):
+    """Valid states for a quantum node.
+
+    This is intended only to capture status to the extent necessary to
+    represent state transitions when executing or retrying quanta; full status
+    information that's useful for diagnostic purposes does not belong here.
+    """
+
+    BUILT = enum.auto()
+    """Quantum has been predicted but is not known to have been started or
+    finished.
+    """
+
+    STARTED = enum.auto()
+    """Execution has started but is not complete.
+
+    Workspace implementations are not required to ever set this state.
+    """
+
+    SUCCEEDED = enum.auto()
+    """Execution completed successfully.
+
+    This includes the case where the quantum had no work to do, which can be
+    determined by looking at the status of downstream dataset nodes.
+    """
+
+    FAILED = enum.auto()
+    """Execution was started but raised an exception or otherwise failed.
+
+    When this status is set on a quantum node, the node may have an
+    ``exception_type`` attribute with the exception type, a `str`
+    ``message`` attribute with the exception message, and a ``traceback``
+    attribute with the exception traceback, *if* an exception was caught.
+    """
+
+
+class DatasetStatus(enum.Flag):
+    """Valid states for a dataset node in quantum graph."""
+
+    PREDICTED = enum.auto()
+    """Dataset has been predicted to exist but is not known to have been
+    written or invalidated.
+    """
+
+    PRESENT = enum.auto()
+    """Dataset has been produced, and the producing quantum was either
+    successful or is still running.
+
+    This is all state for all overall-inputs to the workspace that were
+    present in the central data repository.
+    """
+
+    INVALIDATED = enum.auto()
+    """Dataset has been produced but the producing task later failed.
+
+    This can also occur if this quantum or an upstream one was poisoned
+    (manually marked as failed after it had apparently succeeded).
+
+    Invalidated datasets are not transferred to the central data repository
+    on workspace commit, but they may be reclassified again prior to commit.
+    """
+
+
+class QuantumGraph(ABC):
+    """A directed acyclic graph of datasets and the quanta that produce and
+    consume them.
+
+    Notes
+    -----
+    Datasets are included in a quantum graph if and only if a producing or
+    consuming quantum is included; there are never isolated dataset nodes with
+    no edges in a graph, and any quantum in the graph can be assumed to have
+    all of its nodes.
```

Comment:
Doe it have place for init inputs/outputs? I remember we wanted to have a special quantum producing those.